File: grind75_python/accountsMerge.py
# https://leetcode.com/problems/accounts-merge/description/

from typing import List


class Solution:
  def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
    ans = []
    email_acc = dict()
    visited = [False for _ in range(len(accounts))]

    # consider each name to be an account number
    # create a mapping for email to the account
    for i in range(len(accounts)):
      for j in range(1, len(accounts[i])):
        email = accounts[i][j]
        if email in email_acc:
          email_acc[email].append(i)
        else:
          email_acc[email] = [i]

    def dfs(i, emails):
      visited[i] = True
      for email_idx in range(1, len(accounts[i])):
        email = accounts[i][email_idx]
        emails.add(email)
        for account in email_acc[email]:
          if not visited[account]:
            dfs(account, emails)

    for i in range(len(accounts)):
      if not visited[i]:
        emails = set()
        dfs(i, emails)
        ans.append([accounts[i][0]] + sorted(emails))

    return ans


s = Solution()
print(s.accountsMerge([["John", "johnsmith@mail.com", "john_newyork@mail.com"], ["John",
      "johnsmith@mail.com", "john00@mail.com"], ["Mary", "mary@mail.com"], ["John", "johnnybravo@mail.com"]]))

File: CPP/accountsMerge.cpp
// https://leetcode.com/problems/accounts-merge/

#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  void dfs(unordered_set<string> &visited, unordered_map<string, vector<string>> &adj, string s, vector<string> &vec) {
    visited.insert(s);
    vec.push_back(s);
    for (int i = 0; i < adj[s].size(); i++) {
      if (visited.find(adj[s][i]) == visited.end()) {
        dfs(visited, adj, adj[s][i], vec);
      }
    }
  }
  vector<vector<string>> accountsMerge(vector<vector<string>> &accounts) {
    unordered_map<string, vector<string>> adj;
    unordered_set<string> visited;
    vector<vector<string>> ans;

    for (int i = 0; i < accounts.size(); i++) {
      for (int j = 2; j < accounts[i].size(); j++) {
        adj[accounts[i][1]].push_back(accounts[i][j]);
        adj[accounts[i][j]].push_back(accounts[i][1]);
      }
    }

    for (int i = 0; i < accounts.size(); i++) {
      for (int j = 1; j < accounts[i].size(); j++) {
        if (visited.find(accounts[i][j]) == visited.end()) {
          ans.push_back({accounts[i][0]});
          dfs(visited, adj, accounts[i][j], ans[ans.size() - 1]);
        }
      }
    }

    for (int i = 0; i < ans.size(); i++) {
      sort(++ans[i].begin(), ans[i].end());
    }

    return ans;
  }
};